libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkZstdLibrary.java

```diff
@@ -49,11 +49,15 @@ public long compressBound(int srcLen) {
     }
 
     @Override
-    public long compress(ByteBuffer dst, ByteBuffer src, int compressionLevel) {
-        var nativeDst = MemorySegment.ofBuffer(dst);
-        var nativeSrc = MemorySegment.ofBuffer(src);
+    public long compress(CloseableByteBuffer dst, CloseableByteBuffer src, int compressionLevel) {
+        assert dst instanceof JdkCloseableByteBuffer;
+        assert src instanceof JdkCloseableByteBuffer;
+        var nativeDst = (JdkCloseableByteBuffer) dst;
+        var nativeSrc = (JdkCloseableByteBuffer) src;
+        var segmentDst = nativeDst.segment.asSlice(dst.buffer().position(), dst.buffer().remaining());
+        var segmentSrc = nativeSrc.segment.asSlice(src.buffer().position(), src.buffer().remaining());
         try {
-            return (long) compress$mh.invokeExact(nativeDst, dst.remaining(), nativeSrc, src.remaining(), compressionLevel);
+            return (long) compress$mh.invokeExact(segmentDst, segmentDst.byteSize(), segmentSrc, segmentSrc.byteSize(), compressionLevel);
```

Comment:
Do you need to change the signature of `compress$mh` now that it's taking the src/dst size as a `long` rather than an `int`?

Reply:
I fixed this to pass in an int (the region memory size is wrong anyways)